### PR TITLE
Preserve references to buffers when doing module.apply_

### PR DESCRIPTION
**Pull request by ngimel**
Currently, if one moves a module between devices or casts to a different datatype, by default the references to parameters will be preserved, but references to buffers won't be. This PR makes handling of parameters and buffers consistent.

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -251,7 +252,12 @@ def compute_should_use_set_data(tensor, tensor_applied):
 
         for key, buf in self._buffers.items():
             if buf is not None:
-                self._buffers[key] = fn(buf)
+                buf_applied = fn(buf)
```
Comment (ezyang) [inline torch/nn/modules/module.py:255]: Should this be run under a `no_grad` block?

**Review by ezyang — APPROVED** [review]
Technically bc breaking! Hopefully not in an observable way

File: torch/nn/modules/module.py (hunk 1 above)
Comment (ngimel) [reply torch/nn/modules/module.py:255]: You are right! Buffers almost never require grads, but if they do, this will be an error.

**Comment by ngimel** [comment]
> Technically bc breaking! Hopefully not in an observable way

It is observable :-(
Test failure is real 
```
======================================================================
ERROR: test_NLLLoss_weights_ignore_index_neg_cuda_float (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/home/ngimel/pytorch/torch/testing/_internal/common_utils.py", line 704, in wrapper
    method(*args, **kwargs)
  File "/private/home/ngimel/pytorch/torch/testing/_internal/common_utils.py", line 704, in wrapper
    method(*args, **kwargs)
  File "test_nn.py", line 8442, in <lambda>
    test=test, kwargs=kwargs: test.test_cuda(self, dtype=torch.float, **kwargs))
  File "/private/home/ngimel/pytorch/torch/testing/_internal/common_nn.py", line 4864, in test_cuda
    cpu_output = test_case._forward_criterion(cpu_module, cpu_input, cpu_target, extra_args=extra_args)
  File "test_nn.py", line 299, in _forward_criterion
    output = criterion(input, target, *extra_args)
  File "/private/home/ngimel/pytorch/torch/nn/modules/module.py", line 563, in __call__
    result = self.forward(*input, **kwargs)
  File "/private/home/ngimel/pytorch/torch/nn/modules/loss.py", line 205, in forward
    return F.nll_loss(input, target, weight=self.weight, ignore_index=self.ignore_index, reduction=self.reduction)
  File "/private/home/ngimel/pytorch/torch/nn/functional.py", line 2149, in nll_loss
    ret = torch._C._nn.nll_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index)
RuntimeError: Expected object of device type cuda but got device type cpu for argument #1 'self' in call to _thnn_nll_loss_forward

----------------------------------------------------------------------
```

**Comment by ngimel** [comment]
Ok, this is not going to work without fixing a lot of tests, and I can do what I need to do without it, closing.

**Pull request closed without merging**